pkgs/native_assets_cli/lib/src/validation.dart

```diff
@@ -12,34 +14,58 @@ Future<ValidationErrors> validateBuildInput(BuildInput input) async =>
     _validateHookInput('BuildInput', input);
 
 Future<ValidationErrors> validateLinkInput(LinkInput input) async {
-  final errors = <String>[
+  final recordUses = input.recordedUsagesFile;
+  return <String>[
     ..._validateHookInput('LinkInput', input),
+    if (recordUses != null)
+      ...validateUri(
+        '$LinkInput.recordUses',
+        input.outputDirectoryShared,
+      ),
   ];
-  final recordUses = input.recordedUsagesFile;
-  if (recordUses != null && !File.fromUri(recordUses).existsSync()) {
-    errors.add('LinkInput.recordUses ($recordUses) does not exist.');
-  }
-  return errors;
 }
 
 ValidationErrors _validateHookInput(String inputName, HookInput input) {
+  final errors = <String>[
+    ...validateUri('$inputName.packageRoot', input.packageRoot),
+    ...validateUri('$inputName.outputDirectory', input.outputDirectory),
+    ...validateUri(
+      '$inputName.outputDirectoryShared',
+      input.outputDirectoryShared,
+    ),
+    ...validateUri(
+      '$inputName.outputFile',
+      input.outputFile,
+      mustExist: false,
+    ),
+  ];
+  return errors;
+}
+
+@internal
+ValidationErrors validateUri(
+  String name,
+  Uri uri, {
+  bool mustExist = true,
+  bool mustBeAbsolute = true,
+}) {
   final errors = <String>[];
-  if (!Directory.fromUri(input.packageRoot).existsSync()) {
-    errors.add('$inputName.packageRoot (${input.packageRoot}) '
-        'has to be an existing directory.');
-  }
-  if (!Directory.fromUri(input.outputDirectory).existsSync()) {
-    errors.add('$inputName.outputDirectory (${input.outputDirectory}) '
-        'has to be an existing directory.');
+  if (mustBeAbsolute && !uri.isAbsolute) {
+    errors.add('$name (${uri.toFilePath()}) must be an absolute path.');
   }
-  if (!Directory.fromUri(input.outputDirectoryShared).existsSync()) {
-    errors.add(
-        '$inputName.outputDirectoryShared (${input.outputDirectoryShared}) '
-        'has to be an existing directory');
+  if (mustExist && !_fileSystemEntity(uri).existsSync()) {
+    errors.add('$name (${uri.toFilePath()}) does not exist.');
```

Comment:
Nit: This feels like something which could bite users in the future, when you are confused why a file doesn't exist but it's actually just a directory. Maybe output something like `does not exist as a [file/directory]`?

Reply:
You mean we should check explicitly for whether something is a file or is a directory? And give an error if the protocol mandates a file but the user outputs a directory or vice versa?

Reply:
No, writing to the output what we decided it is is fine.

Reply:
I had already implemented the file vs directory distinction. I think it's a nice addition. E.g. it prevents code failing by passing a file path into a directory and vice versa.